### PR TITLE
feat: booking with phone number

### LINKED ISSUE #14534: [CAL-3403] Allow booking with phone only
Quoting @PeerRich:

# RFC: Allow booking with phone only

A couple people have requested to make bookings without email.

We could offer people this as an option and not send an email/calendar invite, but an SMS instead.

The host can still get a calendar entry (without the guests email, i.e.: [no-reply@CaLxCyMru.com](mailto:no-reply@CaLxCyMru.com)).

When confirming meeting via SMS, we could send a message: “If you would like to receive a calendar invitation, reply with your email.”

After providing an email, we can then update the existing calendar invite for the host

## Mockups

image (2)

image (3)

[Link to original private post](https://app.campsite.co/cal/posts/rv3l714yg8ty)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3403](https://linear.app/calcom/issue/CAL-3403/allow-booking-with-phone-only)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

Fixes  https://github.com/calcom/cal.com/issues/14534
Follow up: https://github.com/calcom/cal.com/issues/15415

How to create event type with only phone number?
1) Create a Team Event type

Screenshot 2024-05-13 at 9 03 09 PM

2) Make Email field not required and toggle off to hide the field

Screenshot 2024-05-13 at 9 03 49 PM

Turn the toggle on for attendee phone number

Screenshot 2024-05-13 at 9 04 07 PM

3) Fill the form 

Screenshot 2024-05-13 at 9 06 16 PM

- Added 4 Unit Tests
   - [Event Type with only Attendee Phone number as required field and Email as hidden field] succesfully creates a booking when the users are available as per the common schedule selected in the event-type
   - [Event Type with both Attendee Phone number and Email as required fields] succesfully creates a booking when the users are available as per the common schedule selected in the event-type
   - [Event Type that requires confirmation with only Attendee Phone number as required field and Email as optional field] succesfully creates a booking when the users are available as per the common schedule selected in the event-type
   - [Event Type with Both Email and Attendee Phone Number as required fields] should send rescheduling emails when round robin is rescheduled to same host

- Added 2 E2E tests
  - Can create a booking for Round Robin EventType with both phone number and email required
  -  Can create a booking for Collective EventType with only phone number

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: apps/web/modules/bookings/views/bookings-single-view.tsx
Comment (joeauyeung) [inline apps/web/modules/bookings/views/bookings-single-view.tsx:356]: If this is unused we should remove this

**Review by joeauyeung — DISMISSED** [review]
Great work @Udit-takkar! This is a pretty big task to tackle so koodos to you. I have a few comments.

## Blocking 
- Have you considered any alternatives to not using the email placeholder if an attendee books via phone number? I had two ideas. Both of these ideas would transform the attendee email column to optional.
	1. Creating an attendee repository and when creating attendees ensuring that the attendee has either an email or a phone number
	2. What if we altered the migration file to add a check constraint to the attendees table. Something like `ADD CONSTRAINT check_contact_info CHECK (email IS NOT NULL OR phone_number IS NOT NULL)`

- We should revaluate the `email-manager` since it is not just sending emails but also SMS messages. We could have a `NotificationManager` class to handle the logic of whether to send emails or SMS messages and to whom

**Comment by Udit-takkar** [comment]
> * We should revaluate the `email-manager` since it is not just sending emails but also SMS messages. We could have a `NotificationManager` class to handle the logic of whether to send emails or SMS messages and to whom

Yeah this is what carina also suggested on my post on campsite. I am planning to add this in my new PR because this one's is already too big.

**Comment by joeauyeung** [comment]
> > Have you considered any alternatives to not using the email placeholder if an attendee books via phone number? I had two ideas. Both of these ideas would transform the attendee email column to optional.
> 
> @joeauyeung Actually making email an optional field would require adding too many checks in the complete codebase and could potentially break some things so peer suggested adding a placeholder "[booked-with-sms@cal.com](mailto:booked-with-sms@cal.com)".
> 
> But yeah for the next version it would best to make the email field optional and adding a constraint like you mentioned. I'll open an issue for this

I would argue we can reduce the amount of code changes we would have to make by having that logic live in an attendee repository class and calling a `createAttendees()` that contains the logic to check for an email or phone number.

**Comment by Udit-takkar** [comment]
> I would argue we can reduce the amount of code changes we would have to make by having that logic live in an attendee repository class and calling a `createAttendees()` that contains the logic to check for an email or phone number.

@joeauyeung  I tried setting email as optional parameter and check constraint but there's just too much code that assumes email is present and a lot of types like `Person` , `TeamMember` would become in compatible. Trying to fix them now

Screenshot 2024-05-15 at 7 42 47 PM

File: packages/features/bookings/lib/getBookingFields.ts
Comment (CarinaWolli) [inline packages/features/bookings/lib/getBookingFields.ts:225]: Better label would be just 'Phone number' instead of 'Attendee Phone Number'

File: packages/sms/sms-manager.ts
Comment (CarinaWolli) [inline packages/sms/sms-manager.ts:21]: we need to add rate limiting here

**Review by CarinaWolli — DISMISSED** [review]
It's not a full review yet, but have some first feedback:

- If email or attendeePhoneNumber is set to hide or optional I would prefer it to automatically change the other field to be required and visible

**Review by CarinaWolli — DISMISSED** [review]
Found some issues when testing: 

- [ ]  When event has workflow with email to attendee we need to automatically add the email booking question
- [x] Seats don't work. When booking the second seat it is throwing ‘Already signed up for this booking’ because all attendees have the same email address we need to check for the phone number instead

- [x]  Phone number shows twice on booking success page

- [x]  Do we want `booked-with-sms@cal.com` as a guest of the calendar event? In the description we should have the attendee’s phone number not the email: 
Screenshot 2024-05-22 at 11 41 06 AM

**Comment by Udit-takkar** [comment]
> Found some issues when testing:
> 
> * [ ]   When event has workflow with email to attendee we need to automatically add the email booking question

Just this one is left which I'll add later. (similar to upsertSmsReminderFieldForBooking)

File: packages/trpc/server/routers/viewer/bookings/requestReschedule.handler.ts
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/bookings/requestReschedule.handler.ts:256]: `team` is not defined on `builder.calendarEvent` so SMS is not sending

Can you check that for all other sendEmailAndSMS functions as well?

**Review by CarinaWolli — DISMISSED** [review]
Left some blocking comments. Also in addition to the requested changes I added, can you check that all SMS are formatted correctly? So, no space before ` . ` No double ` . ` no missing ` . `

File: packages/sms/attendee/event-request-to-reschedule-sms.ts
Comment (CarinaWolli) [reply packages/sms/attendee/event-request-to-reschedule-sms.ts:15]: can we add Reason for reschedule request here?

File: packages/sms/attendee/event-cancelled-sms.ts
Comment (CarinaWolli) [inline packages/sms/attendee/event-cancelled-sms.ts:21]:
```suggestion
```
Booking is canceled and can't be rescheduled or canceled anymore

File: packages/sms/attendee/event-rescheduled-sms.ts
Comment (CarinaWolli) [reply packages/sms/attendee/event-rescheduled-sms.ts:16]: still open, there is already a period at the end of the translation event_type_has_been_rescheduled_on_time_date

File: packages/features/bookings/lib/handleConfirmation.ts
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleConfirmation.ts:107]: this SMS isn't sending because evt.team is undefined

**Review by CarinaWolli — DISMISSED** [review]

**Comment by CarinaWolli** [comment]
I forgot to add that to my review: 

Here are some other issues I found that I think can be handled in a follow up. Could you create issues for them, if not handled in this PR? 
- Two phone number fields when 'SMS to attendee' workflow is enabled on the event type 
- Either change 'requires booker email verification' to 'requires booker sms verification' or disable feature
- rename email-manager (it doesn't only manage emails)
-  (from review above) When event has workflow with email to attendee we need to automatically add the email booking question

**Comment by CarinaWolli** [comment]
@Udit-takkar didn't we talk about making this a org-only feature for now?

File: packages/sms/attendee/event-request-to-reschedule-sms.ts
Comment (CarinaWolli) [inline packages/sms/attendee/event-request-to-reschedule-sms.ts:15]:
```suggestion
    })} \n\n${t("need_to_reschedule_or_cancel")} ${this.calEvent.bookerUrl ?? WEBAPP_URL}/booking/${
```

**Review by CarinaWolli — DISMISSED** [review]
- [x]  I am getting two SMS when rescheduling, it wrongly sends the `sendRoundRobinScheduledEmailsAndSMS` SMS as well

**Comment by Udit-takkar** [comment]
> @Udit-takkar didn't we talk about making this a org-only feature for now?

Yeah. should i wait for a credit system for sending SMS or just make it org only feature ?

**Comment by CarinaWolli** [comment]
> > @Udit-takkar didn't we talk about making this a org-only feature for now?
> 
> Yeah. should i wait for a credit system for sending SMS or just make it org only feature ?

Let's make it org-only for now

**Review by CarinaWolli — DISMISSED** [review]
Great work @Udit-takkar 🙌🏻

**Comment by Udit-takkar** [comment]
@PeerRich Keith suggested that we merge this after retreat.

**Comment by keithwillcode** [comment]
@zomars @emrysal Moving back to draft as we will wait until post-retreat to merge this.

**Comment by PeerRich** [comment]
not stale, lets get this reviewed after merge conflicts are resolved

File: packages/app-store/exchangecalendar/lib/CalendarService.ts
Comment (zomars) [inline packages/app-store/exchangecalendar/lib/CalendarService.ts:66]: Why is the BOOKED_WITH_SMS_EMAIL necessary?

File: packages/features/bookings/lib/handleNewBooking/test/team-bookings/collective-scheduling.test.ts
Comment (zomars) [inline packages/features/bookings/lib/handleNewBooking/test/team-bookings/collective-scheduling.test.ts:944]: Why the change of expectations?

File: packages/platform/atoms/booker/BookerWebWrapper.tsx
Comment (zomars) [inline packages/platform/atoms/booker/BookerWebWrapper.tsx:140]: [Prefer Early Returns](https://handbook.cal.com/engineering/best-practices/prefer-early-returns)

**Review by zomars — CHANGES_REQUESTED** [review]
Overall nice work so far @Udit-takkar. I only have 2 asks to increase the confidence for merging.

- Since this is a significant change of the booking process. Can we at least add a couple of tests cases for our E2E booking flow tests?
- Since email is not required anymore I feel like this is introducing a lot of nesting and overall it's becoming harder to track these fields. What's the rationale behind the DB design? What other alternatives were considered and why was this the winner?

File: apps/web/test/utils/bookingScenario/MockPaymentService.ts
Comment (Udit-takkar) [reply apps/web/test/utils/bookingScenario/MockPaymentService.ts:67]: I have to create a single notification manager class in follow up PR https://github.com/calcom/cal.com/issues/15415#issue-2348287545

**Comment by Udit-takkar** [comment]
> Since this is a significant change of the booking process. Can we at least add a couple of tests cases for our E2E booking flow tests?

Yeah sure. I already few unit tests. is there any specific scenario that i should add?

> Since email is not required anymore I feel like this is introducing a lot of nesting and overall it's becoming harder to track these fields. What's the rationale behind the DB design? What other alternatives were considered and why was this the winner?

The changes in DB schema are:-

I made email an optional field now and added a new field phoneNumber in Attendee model with a constraint that

```
ALTER TABLE "Attendee"
ADD CONSTRAINT check_email_or_phone
CHECK ("email" IS NOT NULL OR "phoneNumber" IS NOT NULL);
```

BOOKED_WITH_SMS is used as a placeholder email address when it's not present as a lot of our codebase depend upon bookerEmail and this is a very big change. Few things like creating a calendar event would also require email of attendee.

Eventually I have to completely remove BOOKED_WITH_SMS after some time on production. This feature is only limited to org team event types for now and we need it fast for cal ai feature which some customers are waiting for.

File: packages/app-store/exchangecalendar/lib/CalendarService.ts
Comment (Udit-takkar) [reply packages/app-store/exchangecalendar/lib/CalendarService.ts:66]: email is required field here to create an event and BOOKED_WITH_SMS would also help us know that it was booked with phone number.

**Comment by zomars** [comment]
> Yeah sure. I already few unit tests. is there any specific scenario that i should add?

Booking with only phone required. Booking with both phone and email required.

**Comment by Udit-takkar** [comment]
> Booking with only phone required. Booking with both phone and email required.

Done.

File: packages/platform/atoms/booker/BookerWebWrapper.tsx
Comment (zomars) [inline packages/platform/atoms/booker/BookerWebWrapper.tsx:140]: Shouldn't this be..
```suggestion
      if (!bookerForm.formEmail) return;
```

?

Comment (Udit-takkar) [reply packages/platform/atoms/booker/BookerWebWrapper.tsx:140]: oh right.

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
I found some places where the SMS wasn't sent: 

- [x] Seated events (SMS is only sent for first seat)
- [x] Location change SMS

@Udit-takkar It's really important that we make sure that we always send a SMS or Email. Can you make sure to test all SMS manually, I might have missed something when testing

**Comment by Udit-takkar** [comment]
> It's really important that we make sure that we always send a SMS or Email. Can you make sure to test all SMS manually, I might have missed something when testing

Fixed and Tested more sms types

1) scheduled seats, rescheduled seats, cancelled seats
2) location changed
3) attendee requested

not adding sms for daily recording, transcript for now.

**Review by CarinaWolli — DISMISSED** [review]
Great work @Udit-takkar! I went over all changes again, and it looks good to me now 👍 The comments are non-blocking

Let's have @zomars give the final approval 🙏

File: packages/emails/email-manager.ts
Comment (Udit-takkar) [reply packages/emails/email-manager.ts:131]: I had test all three functions `sendRoundRobinRescheduledEmailsAndSMS`, `sendRoundRobinScheduledEmailsAndSMS` and `sendRoundRobinCancelledEmailsAndSMS` and we are passing attendees to them in handleNewBooking


```
   const newBookingMemberEmails: Person[] =
          copyEvent.team?.members
            .map((member) => member)
            .concat(copyEvent.organizer)
            .concat(copyEvent.attendees) || [];
            
            ```

File: packages/prisma/extensions/booking-idempotency-key.ts
Comment (Udit-takkar) [reply packages/prisma/extensions/booking-idempotency-key.ts:23]: TS is complaining if i do this  

```
          const isSingleAttendeeCreate = args.data.attendees?.create && !Array.isArray(args.data.attendees?.create);
     if (
            isSingleAttendeeCreate &&
            args.data.attendees?.create.email
          ) {
            uniqueJoinInput.push(args.data.attendees?.create.email);
          }     
          
          
```

**Comment by PeerRich** [comment]
excited for this

**Review by zomars — DISMISSED** [review]
Seems like the schema changes are causing API v2 to fail building:

image

File: packages/app-store/wipemycalother/lib/reschedule.ts
Comment (zomars) [inline packages/app-store/wipemycalother/lib/reschedule.ts:36]: Do we need everything in the eventType? or why are we not using select in here? @Udit-takkar

File: packages/features/bookings/lib/handleNewBooking.ts
Comment (zomars) [reply packages/features/bookings/lib/handleNewBooking.ts:1359]: cc @Udit-takkar

File: packages/platform/atoms/booker/BookerWebWrapper.tsx
Comment (zomars) [inline packages/platform/atoms/booker/BookerWebWrapper.tsx:146]: If we're doing early return above then this should not be neither null nor undefined no? @Udit-takkar 
```suggestion
      verifyEmail.setVerifiedEmail(bookerForm.formEmail);
```

File: packages/trpc/server/routers/viewer/eventTypes/util.ts
Comment (zomars) [inline packages/trpc/server/routers/viewer/eventTypes/util.ts:187]: We don't need `else if` in here since we're throwing and other conditions anyways. 

```suggestion
  if (emailField?.hidden && attendeePhoneNumberField?.hidden) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: `Both Email and Attendee Phone Number cannot be hidden`,
    });
  }
  if (emailField?.hidden && emailField?.required) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: `Email field cannot be hidden if it is a required field. You can click on Edit to mark it as not required field.`,
    });
  }
  if (attendeePhoneNumberField?.hidden && attendeePhoneNumberField?.required) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: `Attendee Phone Number field cannot be hidden if it is a required field. You can click on Edit to mark it as not required field.`,
    });
  }
  if (!emailField?.required && !attendeePhoneNumberField?.required) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: `Atleast Email or Attendee Phone Number need to be required field.`,
    });
  }
```

File: packages/app-store/wipemycalother/lib/reschedule.ts
Comment (Udit-takkar) [reply packages/app-store/wipemycalother/lib/reschedule.ts]: we don't need other event type fields

**Review by joeauyeung — DISMISSED** [review]
Really solid PR @Udit-takkar. When reviewing I was thinking about the long term maintainability of our codebase. We'll now have to start testing against emails & phone numbers. 

- The big thing I noticed is you added a `RequiresAtLeastOne` type. Could we use this in the functions where we see `email?: string; phoneNumber?: string`?
- Thoughts on having a `bookingAttendee` class where we pass the booking attendee. I see the email / phone number as the attendee identifier so we could have a method like `bookingAttendee.getIdentifier()` to get those fields. Rather than calling stuff like getEmail or getPhoneNumber? Could also provide some safety if the email or phone number doesn't exist.
- I'm seeing two instances of `getEventTypesFromDB` can we stack a PR on top of this to remove the two instances?
- Could we create a prisma middleware when creating an attendee to ensure an email or phone number is present?

**Comment by zomars** [comment]
> Really solid PR @Udit-takkar. When reviewing I was thinking about the long term maintainability of our codebase. We'll now have to start testing against emails & phone numbers.
> 
> * The big thing I noticed is you added a `RequiresAtLeastOne` type. Could we use this in the functions where we see `email?: string; phoneNumber?: string`?
> * Thoughts on having a `bookingAttendee` class where we pass the booking attendee. I see the email / phone number as the attendee identifier so we could have a method like `bookingAttendee.getIdentifier()` to get those fields. Rather than calling stuff like getEmail or getPhoneNumber? Could also provide some safety if the email or phone number doesn't exist.
> * I'm seeing two instances of `getEventTypesFromDB` can we stack a PR on top of this to remove the two instances?
> * Could we create a prisma middleware when creating an attendee to ensure an email or phone number is present?

Pretty valid points. @joeauyeung I was also thinking something similar. We could even treat the email field as a phone number input in order to avoid changing something as fundamental as email in our booking process. 

This is not a "simple" change at all.

I'd rather use phones as as "faux email" so you just type you phone number: +1 555 555 5555 but internally it gets handled as a email: `15555555555-phone@cal.com` or something like that.

On the technical side, treating phone numbers as "faux emails" could simplify how the system works behind the scenes. By handling all identifiers in a similar way, you might reduce the need for complicated changes to the system, keeping things running smoothly without a major overhaul. It’s a way to make the switch without disrupting everything.

**Comment by PeerRich** [comment]
maybe 15555555555@sms.cal.com?

**Review by tarunkumarmetaforms — APPROVED** [review]

**Review by joeauyeung — DISMISSED** [review]
I think we still need to decide how we're going to handle maintaining booking with phone numbers & our code base before giving an approval. I do have some comments around the functionality

## Blocking
- [ ] On the organizer calendar event the placeholder email is appearing as an attendee
## Non-blocking
- [x] When editing the booking fields, even if a field is hidden the required field check is still run. This might could be annoying when transitioning from email -> phone numbers for bookings
- [x] I'm consistently seeing a missing padding between the organizer and attendee in the calendar event description 
image

**Comment by Udit-takkar** [comment]
> On the organizer calendar event the placeholder email is appearing as an attendee

Do you mean here ?

Screenshot 2024-08-26 at 3 24 31 PM

We need some email to create a calendar event. Do you suggest we don't add any email in attendee when email is not passed?

**Comment by Udit-takkar** [comment]
> When editing the booking fields, even if a field is hidden the required field check is still run. This might could be annoying when transitioning from email -> phone numbers for bookings

Okay. now I changed the logic.

**Comment by joeauyeung** [comment]
> > On the organizer calendar event the placeholder email is appearing as an attendee
> 
> Do you mean here ?
> 
> Screenshot 2024-08-26 at 3 24 31 PM
> 
> We need some email to create a calendar event. Do you suggest we don't add any email in guests when email is not passed?

I thought we didn't want to the organizers to see the placeholder emails. If we're ok with that then I'm ok. We just have to remember to create a catch all rule so organizers don't see this email from their calendar

image

**Comment by zomars** [comment]
Fixing conflicts before testing.

File: packages/features/bookings/lib/handleNewBooking.ts
Comment (zomars) [reply packages/features/bookings/lib/handleNewBooking.ts:1354]: @alishaz-polymath @Udit-takkar can we resolve this?

File: packages/lib/isSmsCalEmail.ts
Comment (zomars) [inline packages/lib/isSmsCalEmail.ts:1]: Can anyone put these kind of emails at the email field when booking? Would that introduce unintended behavior? @Udit-takkar @alishaz-polymath

**Review by zomars — APPROVED** [review]

**Pull request merged**
